memory/src/main/scala/filodb.memory/format/vectors/HistogramVector.scala

```diff
@@ -161,25 +161,28 @@ object HistogramVector {
   val OffsetNumBuckets = 11
   // After the bucket area are regions for storing the counter values or pointers to them
 
-  final def getNumBuckets(addr: Ptr.U8): Int = addr.add(OffsetNumBuckets).asU16.getU16
+  final def getNumBuckets(acc: MemoryReader, addr: Ptr.U8): Int = addr.add(OffsetNumBuckets).asU16.getU16(acc)
 
-  final def getNumHistograms(addr: Ptr.U8): Int = addr.add(OffsetNumHistograms).asU16.getU16
-  final def resetNumHistograms(addr: Ptr.U8): Unit = addr.add(OffsetNumHistograms).asU16.asMut.set(0)
-  final def incrNumHistograms(addr: Ptr.U8): Unit =
-    addr.add(OffsetNumHistograms).asU16.asMut.set(getNumHistograms(addr) + 1)
+  final def getNumHistograms(acc: MemoryReader, addr: Ptr.U8): Int = addr.add(OffsetNumHistograms).asU16.getU16(acc)
+  final def resetNumHistograms(acc: MemoryAccessor, addr: Ptr.U8): Unit =
+    addr.add(OffsetNumHistograms).asU16.asMut.set(acc, 0)
+  final def incrNumHistograms(acc: MemoryAccessor, addr: Ptr.U8): Unit =
+    addr.add(OffsetNumHistograms).asU16.asMut.set(acc, getNumHistograms(acc, addr) + 1)
 
   // Note: the format code defines bucket definition format + format of each individual compressed histogram
-  final def formatCode(addr: Ptr.U8): Byte = addr.add(OffsetFormatCode).getU8.toByte
-  final def afterBucketDefAddr(addr: Ptr.U8): Ptr.U8 = addr + OffsetBucketDef + bucketDefNumBytes(addr)
-  final def bucketDefNumBytes(addr: Ptr.U8): Int = addr.add(OffsetBucketDefSize).asU16.getU16
+  final def formatCode(acc: MemoryReader, addr: Ptr.U8): Byte = addr.add(OffsetFormatCode).getU8(acc).toByte
+  final def afterBucketDefAddr(acc: MemoryReader, addr: Ptr.U8): Ptr.U8 =
+    addr + OffsetBucketDef + bucketDefNumBytes(acc, addr)
+  final def bucketDefNumBytes(acc: MemoryReader, addr: Ptr.U8): Int =
+    addr.add(OffsetBucketDefSize).asU16.getU16(acc)
   final def bucketDefAddr(addr: Ptr.U8): Ptr.U8 = addr + OffsetBucketDef
 
   // Matches the bucket definition whose # bytes is at (base, offset)
-  final def matchBucketDef(hist: BinaryHistogram.BinHistogram, addr: Ptr.U8): Boolean =
-    (hist.formatCode == formatCode(addr)) &&
-    (hist.bucketDefNumBytes == bucketDefNumBytes(addr)) && {
-      UnsafeUtils.equate(UnsafeUtils.ZeroPointer, bucketDefAddr(addr).addr, hist.buf.byteArray, hist.bucketDefOffset,
-                         hist.bucketDefNumBytes)
+  final def matchBucketDef(hist: BinaryHistogram.BinHistogram, acc: MemoryReader, addr: Ptr.U8): Boolean =
+    (hist.formatCode == formatCode(acc, addr)) &&
+    (hist.bucketDefNumBytes == bucketDefNumBytes(acc, addr)) && {
+      UnsafeUtils.equate(acc.base, bucketDefAddr(addr).addr,
```

Comment:
Again, I think you need `acc.baseOffset + bucketDefAddr(....)` here.

Reply:
Good catch